doc/nvgre_tunnel/nvgre_tunnel.md

```diff
@@ -0,0 +1,448 @@
+             / [ *6( h16 ":" ) h16 ] "::"
+```
+
+```
+; Defines schema for NVGRE Tunnel map configuration attributes
+key                                   = NVGRE_TUNNEL|tunnel_name|vlan_name ; NVGRE tunnel configuration
+; field                               = value
+vsid                                  = DIGITS                             ; 1 to 16 million values
+vlan                                  = 1\*4DIGIT                          ; 1 to 4094 Vlan id
+```
+
+#### Orchestration agent
+
+The following orchestration agents will be added or modified. The flow diagrams are captured in a later section.
+
+##### Figure 1. Orchestration agents
+
+<p align=center>
+<img src="images/nvgre_orch.svg" alt="Figure 1. Orchestration agents">
+</p>
+
+#### NvgreOrch
+
+`nvgreorch` - it is an orchestration agent that handles the configuration requests directly from ConfigDB. The `nvgreorch` is responsible for creating the tunnel and attaches encap and decap mappers. Both VLAN and Bridge tunnel maps are created for NVGRE tunnel by default, so it would be possible to apply both VLAN/VSID and Bridge/VSID map entries.
+
+#### Orchdaemon
+
+`orchdaemon` - it is the main orchestration agent, which handles all Redis DB's updates, then calls appropriate orchagent, the new `nvgreorch` should be registered inside an `orchdaemon`.
+
+### High-Level Design 
+
+The following sub-modules will be modified:
+* [sonic-swss](https://github.com/Azure/sonic-swss) - will be extended with the new orchestration agent for NVGRE.
+* [sonic-swss-common](https://github.com/Azure/sonic-swss-common) - will be extended with the new tables for ConfigDB.
+* [sonic-utilities](https://github.com/Azure/sonic-utilities) - will be extened with the new CLI.
+
+#### Sequence diagrams
+
+##### Figure 2. NVGRE Tunnel creation flow
+
+<p align=center>
+<img src="images/nvgre_tunnel_create_uml.svg" alt="Figure 2. NVGRE Tunnel creation flow">
+</p>
+
+##### Figure 3. NVGRE Tunnel Map creation flow
+
+<p align=center>
+<img src="images/nvgre_tunnel_map_create_uml.svg" alt="Figure 3. NVGRE Tunnel Map creation flow">
+</p>
+
+##### Figure 4. NVGRE Tunnel delete flow
+
+Note that deleting a tunnel requires the user to remove all the configuration that are associated with this tunnel, it will not be done for the user automatically.
+
+<p align=center>
+<img src="images/nvgre_tunnel_delete_uml.svg" alt="Figure 4. NVGRE Tunnel delete flow">
+</p>
+
+##### Figure 5. NVGRE Tunnel Map delete flow
+
+<p align=center>
+<img src="images/nvgre_tunnel_map_delete_uml.svg" alt="Figure 5. NVGRE Tunnel Map delete flow">
+</p>
+
+##### Figure 6. NVGRE Tunnel CLI config
+
+<p align=center>
+<img src="images/nvgre_cli_config.svg" alt="Figure 6. NVGRE Tunnel CLI config">
+</p>
+
+##### Figure 7. NVGRE Tunnel CLI show
+
+<p align=center>
+<img src="images/nvgre_cli_show.svg" alt="Figure 7. NVGRE Tunnel CLI show">
+</p>
+
+### SAI API 
+
+The NVGRE Tunnel feature require at least `SAI 1.9` or above.
+
```

Comment:
Some aspects of the SAI API in the initial implementation are not described in the HLD. They should be described:

- SAI_TUNNEL_ATTR_PEER_MODE_P2MP (default)
- Only decap mapper entries are created, not encap mapper entries

More importantly, the implementation does not conform to normal SAI tunnel usage. While the implementation seems to be focused on decap, there does not seem to be any code creating sai_tunnel_term_table entries. As far as I know the SAI tunnel definitions, I do not believe any packets will be consumed without entries in the sai_tunnel_term_table. Just creation of sai_tunnel objects is not enough.

Reply:
Added the tunnel termination entry type and fixed implementation.
Remove the encap mappers.

Reply:
@svshah-intel Please review as well. Thanks.